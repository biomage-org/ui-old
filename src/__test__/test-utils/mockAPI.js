import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

// A ticket has been created to address this : https://biomage.atlassian.net/browse/BIOMAGE-1553
import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => Promise.resolve({
  status: code,
  body: JSON.stringify(body),
});

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerResponse = (body) => promiseResponse(body);

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(
    Object.keys(apiMapping),
    (urlStub) => path.endsWith(urlStub),
  );

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  workerResponse,
  delayedResponse,
};
