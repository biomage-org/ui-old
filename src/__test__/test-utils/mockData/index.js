import _ from 'lodash';
import generateMockSamples from './generateMockSamples';
import generateMockExperiments from './generateMockExperiments';
import generateMockProcessingConfig from './generateMockProcessingConfig';

// This file is a work in progress to generate mock data for the endpoints of the test.
// The generated data should be usable as response data to be sent via the mocked API in mockAPI.js
// The module should also expose the generated data that can be imported and used easily in tests
// A ticket has been created to give more thought into this https://biomage.atlassian.net/browse/BIOMAGE-1553

const responseData = {};

responseData.experiments = generateMockExperiments(2);

responseData.samples = [generateMockSamples(
  responseData.experiments[0],
  3,
)];

responseData.processingConfig = generateMockProcessingConfig(3);

// Add samples to first experiment
const samples = responseData.samples[0];
const sampleIds = _.map(samples, 'id');
responseData.experiments[0].samplesOrder = sampleIds;

const { experiments, processingConfig } = responseData;

export {
  experiments,
  samples,
  processingConfig,
  responseData,
};
