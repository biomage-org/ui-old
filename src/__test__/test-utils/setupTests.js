import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
