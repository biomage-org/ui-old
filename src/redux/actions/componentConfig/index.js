import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import getDotPlot from 'redux/actions/componentConfig/getDotPlot';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  getDotPlot,
  updatePlotData,
  resetPlotConfig,
};
