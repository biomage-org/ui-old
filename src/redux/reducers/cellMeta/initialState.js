const mitochondrialContent = {
  loading: true,
  error: false,
  data: [],
};

const doubletScores = {
  loading: true,
  error: false,
  data: [],
};

const numOfGenes = {
  loading: true,
  error: false,
  data: [],
};

const numOfUmis = {
  loading: true,
  error: false,
  data: [],
};

const initialState = {
  mitochondrialContent,
  doubletScores,
  numOfGenes,
  numOfUmis,
};

export default initialState;
