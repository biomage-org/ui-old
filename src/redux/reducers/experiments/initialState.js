const experimentTemplate = {
  name: null,
  description: 'Analysis description',
  id: null,
  createdAt: null,
  updatedAt: null,
  notifyByEmail: true,
  meta: { organism: null, type: '10x' },
  sampleIds: [],
  metadataKeys: [],
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
    activeExperimentId: null,
  },
  ids: [],
};

const METADATA_DEFAULT_VALUE = 'N.A.';

export default initialState;
export { experimentTemplate, METADATA_DEFAULT_VALUE };
