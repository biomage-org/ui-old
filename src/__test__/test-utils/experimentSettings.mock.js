import initialState from 'redux/reducers/experimentSettings/initialState';

const sampleifiedConfig = (sampleIds, configToReplicate) => {
  const result = sampleIds.reduce(
    (acum, sampleId) => {
      // eslint-disable-next-line no-param-reassign
      acum[sampleId] = configToReplicate;
      return acum;
    },
    {},
  );

  return result;
};

const generateProcessingConfigMock = (sampleIds) => ({
  classifier: {
    enabled: true,
    prefiltred: false,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        FDR: 0.1,
      },
    }),
  },
  cellSizeDistribution: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        minCellSize: 10800,
        binStep: 200,
      },
    }),
  },
  mitochondrialContent: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        method: 'absoluteThreshold',
        methodSettings: {
          absoluteThreshold: {
            maxFraction: 0.1,
            binStep: 200,
          },
        },
      },
    }),
  },
  numGenesVsNumUmis: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        regressionType: 'gam',
        regressionTypeSettings: {
          gam: {
            'p.level': 0.00009,
          },
        },
      },
    }),
  },
  doubletScores: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        probabilityThreshold: 0.2,
        binStep: 0.05,
      },
    }),
  },
  dataIntegration: {
    enabled: true,
    dataIntegration: {
      method: 'seuratv4',
      methodSettings: {
        seuratv4: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        fastmnn: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        unisample: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
      },
    },
    dimensionalityReduction: {
      method: 'rpca',
      numPCs: 30,
      excludeGeneCategories: [],
    },
  },
  configureEmbedding: {
    enabled: true,
    embeddingSettings: {
      method: 'umap',
      methodSettings: {
        umap: {
          minimumDistance: 0.1,
          distanceMetric: 'euclidean',
        },
        tsne: {
          perplexity: 30,
          learningRate: 200,
        },
      },
    },
    clusteringSettings: {
      method: 'louvain',
      methodSettings: {
        louvain: {
          resolution: 0.5,
        },
      },
    },
  },
});

const generateExperimentSettingsMock = (sampleIds) => {
  const mockedProcessingConfig = {
    ...initialState.processing,
    ...generateProcessingConfigMock(sampleIds),
  };

  return {
    ...initialState,
    processing: mockedProcessingConfig,
    originalProcessing: mockedProcessingConfig,
    info: {
      ...initialState.info,
      pipelineVersion: 1,
    },
  };
};

export default generateExperimentSettingsMock;
