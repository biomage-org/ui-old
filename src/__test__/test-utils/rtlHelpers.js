import { fireEvent, screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

const selectOption = async (optionName, select) => {
  await userEvent.click(select);
  // userEvent click doesnt trigger the onChange in select
  fireEvent.click(screen.getByTitle(optionName));
};

// dropFilesIntoDropzone takes an input element and simulates droping files into it
// inputElement: this needs to be the actual input element not the dropzone itself, e.g.:
// * <input data-testid='drop-input'/> => screen.getByTestId('drop-input');
// files should be a list of files, e.g.:
// * new File(['content'], 'file.txt', { type: 'text/plain' });
const dropFilesIntoDropzone = async (inputElement, files) => {
  await act(async () => {
    fireEvent.drop(inputElement, {
      target: { files },
    });
  });
};

export {
  // eslint-disable-next-line import/prefer-default-export
  selectOption,
  dropFilesIntoDropzone,
};
