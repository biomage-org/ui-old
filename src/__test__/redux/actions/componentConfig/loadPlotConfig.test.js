import { act } from 'react-dom/test-utils';
import _ from 'lodash';

import '__test__/test-utils/mockWorkerBackend';

import { makeStore } from 'redux/store';

import pushNotificationMessage from 'utils/pushNotificationMessage';
import { loadPlotConfig } from 'redux/actions/componentConfig';
import { initialPlotConfigStates } from 'redux/reducers/componentConfig/initialState';

import endUserMessages from 'utils/endUserMessages';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import fake from '__test__/test-utils/constants';

jest.mock('utils/pushNotificationMessage');

let testStore = null;

enableFetchMocks();

const experimentId = fake.EXPERIMENT_ID;
const plotUuid = 'cellSizeDistributionHistogramMain';
const plotType = 'cellSizeDistributionHistogram';

const mockData = {
  plotData: [1, 2, 3, 4],
  config: {},
};

const mockConfigData = _.merge({}, initialPlotConfigStates[plotType], mockData);

describe('loadPlotConfig', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
    fetchMock.resetMocks();
    testStore = makeStore();
  });

  it('Loads config properly', async () => {
    fetchMock.mockResponse(() => Promise.resolve(new Response(JSON.stringify(mockConfigData))));

    await act(async () => {
      await testStore.dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
    });

    // Hits the correct url
    const fetchUrl = fetchMock.mock.calls[0][0];
    expect(fetchUrl).toEqual(`http://localhost:3000/v2/experiments/${experimentId}/plots/${plotUuid}`);

    // Expect componentConfig to contain key for plotUuid
    expect(testStore.getState().componentConfig[plotUuid]).toBeDefined();

    const plotConfig = testStore.getState().componentConfig[plotUuid];

    expect(plotConfig).toMatchSnapshot();
  });

  it('Loads default initial config for plots if plot config is not found', async () => {
    fetchMock.mockResponse(() => Promise.resolve({ status: 404, body: JSON.stringify('Plot config not found') }));

    await act(async () => {
      await testStore.dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
    });

    // Expect componentConfig to contain key for plotUuid
    expect(testStore.getState().componentConfig[plotUuid]).toBeDefined();

    const plotConfig = testStore.getState().componentConfig[plotUuid];

    expect(plotConfig).toMatchSnapshot();
  });

  it('Non-200 response shows an error message', async () => {
    fetchMock.mockResponse(() => Promise.resolve({ status: 500, body: JSON.stringify('Server error') }));

    await act(async () => {
      await testStore.dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
    });

    // Expect componentConfig to contain key for plotUuid
    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith('error', endUserMessages.ERROR_FETCHING_PLOT_CONFIG);
  });

  it('Invalid response format show an error notification', async () => {
    fetchMock.mockResponse(() => Promise.resolve(new Response('a text body')));

    await act(async () => {
      await testStore.dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
    });

    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith('error', expect.anything());
  });
});
