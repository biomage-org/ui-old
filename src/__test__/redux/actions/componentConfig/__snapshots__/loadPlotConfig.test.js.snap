// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadPlotConfig Loads config properly 1`] = `
Object {
  "config": Object {
    "axes": Object {
      "domainWidth": 2,
      "gridOpacity": 0,
      "gridWidth": 10,
      "labelFontSize": 12,
      "offset": 0,
      "titleFontSize": 13,
      "xAxisRotateLabels": false,
      "xAxisText": "#UMIs in cell",
      "yAxisText": "#UMIs * #Cells",
    },
    "axesRanges": Object {
      "xAxisAuto": true,
      "xMax": 10,
      "xMin": 0,
      "yAxisAuto": true,
      "yMax": 10,
      "yMin": 0,
    },
    "colour": Object {
      "gradient": "default",
      "invert": "standard",
      "masterColour": "#000000",
      "reverseColourBar": false,
      "toggleInvert": "#FFFFFF",
    },
    "dimensions": Object {
      "height": 400,
      "width": 530,
    },
    "fontStyle": Object {
      "colour": "#000000",
      "font": "sans-serif",
    },
    "labels": Object {
      "enabled": true,
      "size": 18,
    },
    "legend": Object {
      "colour": "#000000",
      "enabled": true,
      "position": "top-left",
    },
    "marker": Object {
      "opacity": 5,
      "shape": "circle",
      "showOpacity": true,
      "size": 5,
    },
    "minCellSize": 10800,
    "spec": "1.0.0",
    "title": Object {
      "anchor": "start",
      "dx": 10,
      "fontSize": 20,
      "text": "",
    },
  },
  "error": false,
  "experimentId": "testae48e318dab9a1bd0bexperiment",
  "loading": false,
  "outstandingChanges": false,
  "plotData": Array [
    1,
    2,
    3,
    4,
  ],
  "plotType": "cellSizeDistributionHistogram",
}
`;

exports[`loadPlotConfig Loads default initial config for plots if plot config is not found 1`] = `
Object {
  "config": Object {
    "axes": Object {
      "domainWidth": 2,
      "gridOpacity": 0,
      "gridWidth": 10,
      "labelFontSize": 12,
      "offset": 0,
      "titleFontSize": 13,
      "xAxisRotateLabels": false,
      "xAxisText": "#UMIs in cell",
      "yAxisText": "#UMIs * #Cells",
    },
    "axesRanges": Object {
      "xAxisAuto": true,
      "xMax": 10,
      "xMin": 0,
      "yAxisAuto": true,
      "yMax": 10,
      "yMin": 0,
    },
    "colour": Object {
      "gradient": "default",
      "invert": "standard",
      "masterColour": "#000000",
      "reverseColourBar": false,
      "toggleInvert": "#FFFFFF",
    },
    "dimensions": Object {
      "height": 400,
      "width": 530,
    },
    "fontStyle": Object {
      "colour": "#000000",
      "font": "sans-serif",
    },
    "labels": Object {
      "enabled": true,
      "size": 18,
    },
    "legend": Object {
      "colour": "#000000",
      "enabled": true,
      "position": "top-left",
    },
    "marker": Object {
      "opacity": 5,
      "shape": "circle",
      "showOpacity": true,
      "size": 5,
    },
    "minCellSize": 10800,
    "spec": "1.0.0",
    "title": Object {
      "anchor": "start",
      "dx": 10,
      "fontSize": 20,
      "text": "",
    },
  },
  "error": false,
  "experimentId": "testae48e318dab9a1bd0bexperiment",
  "loading": false,
  "outstandingChanges": false,
  "plotData": Array [],
  "plotType": "cellSizeDistributionHistogram",
}
`;
